enterprise/cluster/src/test/java/org/neo4j/cluster/FreePorts.java

```diff
@@ -0,0 +1,165 @@
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.cluster;
+
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Utility for discovering free ports within a given range, also for remembering already discovered ports
+ * so that next allocation will be faster. This utility is designed to be shared among many tests,
+ * preferably all tests in an entire component.
+ *
+ * One test should call {@link #newSession()} and allocate its ports there and {@link Session#close() close}
+ * when test is completed. Sessions share discovered ports for faster port allocation.
+ */
+public class FreePorts
+{
+    private static final int NOT_FOUND = -1;
+
+    private final Set<Integer> discoveredPorts = new HashSet<>();
+    private final Set<Integer> deniedPorts = new HashSet<>();
+    private final Set<Integer> currentlyOccupiedPorts = new HashSet<>();
+
+    public Session newSession()
+    {
+        return new Session();
+    }
+
+    private synchronized int findFreePort( int minPort, int maxPort ) throws IOException
+    {
+        int port = findFreeAlreadyDiscoveredPort( minPort, maxPort );
+        if ( port == NOT_FOUND )
+        {
+            port = discoverPort( minPort, maxPort );
+        }
+
+        currentlyOccupiedPorts.add( port );
+        return port;
+    }
+
+    private synchronized void releasePort( int port )
+    {
+        if ( !currentlyOccupiedPorts.remove( port ) )
+        {
+            throw new IllegalStateException( "Port " + port + " not occupied" );
+        }
+    }
+
+    private int discoverPort( int minPort, int maxPort ) throws IOException
+    {
+        int port;
+        for ( port = minPort; port <= maxPort; port++ )
+        {
+            if ( discoveredPorts.contains( port ) || deniedPorts.contains( port ) )
+            {
+                // This port has already been discovered or denied
+                continue;
+            }
+            try
+            {
+                // try binding it at wildcard
+                ServerSocket ss = new ServerSocket();
+                ss.setReuseAddress( false );
+                ss.bind( new InetSocketAddress( port ) );
+                ss.close();
+            }
+            catch ( IOException e )
+            {
+                deniedPorts.add( port );
```

Comment:
I didn't want to complicate the PR by changing how free ports are discovered since new failures would be more difficult to pin point. Knowing that this is merely a test utility I'd opt for not changing unless we need to.